components/browser_kit/introduction.rst

```diff
@@ -0,0 +1,233 @@
+.. index::
+   single: BrowserKit
+   single: Components; BrowserKit
+
+The BrowserKit Component
+========================
+
+    The BrowserKit component simulates the behavior of a web browser, allowing
+    you to make requests, click on links and submit forms programmatically.
+
+Installation
+------------
+
+You can install the component in two different ways:
+
+* :doc:`Install it via Composer </components/using_components>`
+  (``symfony/browser-kit`` on `Packagist`_);
+* Use the official Git repository (https://github.com/symfony/BrowserKit).
+
+Basic Usage
+-----------
+
+Creating a Client
+~~~~~~~~~~~~~~~~~
+
+The component only provides an abstract client and does not provide any backend
+ready to use for the HTTP layer.
+
+To create your own client you must extend the abstract client class and
+implement the :method:`Symfony\\Component\\BrowserKit\\Client::doRequest` method.
+This method accepts a request and should return a response::
+
+    namespace Acme;
+
+    use Symfony\Component\BrowserKit\Client as BaseClient;
+    use Symfony\Component\BrowserKit\Response;
+
+    class Client extends BaseClient
+    {
+        protected function doRequest($request)
+        {
+            // convert request into a response
+            // ...
+
+            return new Response($content, $status, $headers);
+        }
+    }
+
+For a simple implementation of a browser based on an HTTP layer, have a look
+at `Goutte`_. For an implementation based on ``HttpKernelInterface``, have a
+look at the Client provided by the :doc:`/components/http_kernel/introduction`.
+
+Making Requests
+~~~~~~~~~~~~~~~
+
+Use the :method:`Symfony\\Component\\BrowserKit\\Client::request` method to make
+any HTTP request. The first two arguments are for the HTTP method and the
+requested URL::
+
+    use Acme\Client;
+
+    $client = new Client();
+    $crawler = $client->request('GET', 'http://symfony.com');
+
+The value returned by the ``request()`` method is an instance of the
+:class:`Symfony\\Component\\DomCrawler\\Crawler` class, which allows accessing
+and traversing HTML elements programmatically.
+
+Clicking Links
+~~~~~~~~~~~~~~
+
+The ``Crawler`` object is capable of simulating link clicks. First, pass the
+text content of the link to the ``selectLink()`` method, which returns you a
+``Link`` object. Then, pass this object to the ``click()`` method, which makes
+the needed HTTP GET request to simulate the link click::
+
+    use Acme\Client;
+
+    $client = new Client();
+    $crawler = $client->request('GET', 'http://symfony.com');
+    $link = $crawler->selectLink('Go elsewhere...')->link();
+    $client->click($link);
+
+Submitting Forms
+~~~~~~~~~~~~~~~~
+
+The ``Crawler`` object is also capable of simulating form submissions. First,
```

Comment:
The client object is capable of simulating form submissions, Crawler is only capable of selecting form elements (and changing the value's).